### DISCUSSION TITLE
Conflict between "noplainbuffer" and "camel" in spelloptions

### DISCUSSION THREAD

**Discussion by ynhhoJ** [neovim/neovim#24643] [Q&A] [answered]
### Problem

I currently utilize two languages — English and Romanian — while having the `spelloptions = "camel"` setting enabled in my config.

Nevertheless, a recurring source of frustration arises when I open any file, only to find that the spellcheck verification is inactive. This compels me to input `:set spelloptions=camel` manually each time I wish to activate the spellchecker.

After some time, i found what `noplainbuffer` is appended when treesitter is trriggered: https://github.com/neovim/neovim/blob/72e619ca92b527140e1323f4e281b7f3ad35f380/runtime/lua/vim/treesitter/highlighter.lua#L134

<details>
  <summary>How it looks now</summary>

![image](https://github.com/neovim/neovim/assets/22500212/84f1848e-7820-4368-a3fe-02c2a8ae7b63)

![image](https://github.com/nvim-treesitter/nvim-treesitter/assets/22500212/22a0fa21-efd2-481a-a528-8cc7d9d77960)
 </details>

<details>
  <summary>How i expect to be</summary>

![image](https://github.com/neovim/neovim/assets/22500212/65b03b51-601d-4111-99fe-0b3cc7179837)

![image](https://github.com/neovim/neovim/assets/22500212/e2d6fc82-3ad4-48cd-aa85-71c8dd950abb)
 </details>

Neovim options:

```lua
local options = {
  relativenumber = true,
  spell = true,
  spelllang = "en_us,ro",
  spelloptions = "camel",
}

for k, v in pairs(options) do
  vim.opt[k] = v
end
```

<details>
<summary>Test code</summary>

```js
console.log("Aplle");
console.log("Bananna");
console.log("Caffine");
console.log("Diamound");
console.log("Elixer");
console.log("Febuary");
console.log("Gravvy");
console.log("Habbit");
console.log("Incorect");
console.log("Jurney");
console.log("Kitchin");
console.log("Latttuce");
console.log("Mischevious");
console.log("Nise");
console.log("Occured");
console.log("Pannini");
console.log("Qeustion");
console.log("Relevent");
console.log("Seperete");
console.log("Truely");
console.log("Unkown");
console.log("Vaccum");
console.log("Wendsday");
console.log("Xceitement");
console.log("Yerning");

console.log('applePie')
console.log('BananaSplit')
console.log('ChocolateCake')
console.log('dougnutHole')
console.log('ElephantEars')
console.log('Frappucino')
console.log('GrapfruiteJuice')
console.log('HamburgrHelper')
console.log('iceCreem')
console.log('JalepenoPopper')
console.log('KeyLimePie')
console.log('LemonMerrinque')
console.log('Marshmellow')
console.log('NoodleSopp')
console.log('OnionRings')
console.log('PanniniPress')
console.log('Quessadilla')
console.log('Raspberrry')
console.log('StrawberrieShortcake')
console.log('Tiramisu')
console.log('UnicorneCake')
console.log('VanillaBean')
console.log('WaffleCone')
console.log('XtraFudge')
console.log('Yogourt')
```
</details>

### Expected behavior

Option `noplainbuffer` should not block spellcheck verification, even if `camel` is enabled or `noplainbuffer` should not be added to `spelloptions` if `camel` option already exists

**Comment by zeertzjq** [comment]
How does this have anything to do with `camel`?

**Comment by lewis6991** [comment]
Is this because treesitter doesn't spell check strings?

I don't think there is an issue here.

**Reply by clason** [reply]
I suspect that is the case, yes. Without a full reproducing config, we're left at guessing, though.

**Comment by ynhhoJ** [comment]
> How does this have anything to do with `camel`?

In my case, spellcheck works only when i set `camel` to `spelloptions`

**Comment by lewis6991** [comment]
When you remove `noplainbuffer`, you're telling Neovim to spellcheck all text, and not the regions that treesitter specifies.

You can check if a node is marked for spellchecking by placing your cursor over some text and running `:Inspect` and check if it includes a `@spell` capture.

**Comment by ynhhoJ** [comment]
@lewis6991 
Tried to place my cursor to `Aplle` text and execute `:Inspect`. 
Result: 
<details>

![image](https://github.com/neovim/neovim/assets/22500212/77fc87ca-1345-43ae-8999-cc55aff7dd93)
</details>

Is there a way to see spellcheck result (?) of specified regions by treesitter?
I don't understand how it works :(

**Reply by lewis6991** [reply, marked as answer]
You can add a file to `$XDG_CONFIG_HOME/nvim/queries/javascript/highlights.scm` with the content:

```scheme
;; extends
(string) @spell
```

And that should enable spellchecking for strings in javascript files.

Currently there is no way to enable spellchecking of strings in all filetypes but it is planned.

**Reply by clason** [reply]
Yes. If the location under the cursor is meant to be spell-checked, you'd see something like
```
  - @spell.javascript links to @spell javascript
```

**Reply by ynhhoJ** [reply]
Thank you for help!
After adding `nvim/queries/javascript/highlights.scm` file, it now looks like this:

![image](https://github.com/neovim/neovim/assets/22500212/9ba991b5-40b9-4e97-8a40-296580636574)

**Comment by sajadspeed** [comment]
If you use `lazy.nvim`:
``` lua
local function set_spell_settings()
	vim.opt_local.spelloptions = "camel"
	vim.opt_local.spellcapcheck = ""
end

vim.api.nvim_create_autocmd("User", {
	pattern = "VeryLazy",
	callback = function()
		-- For first buffer load if exist.
		set_spell_settings()

		vim.api.nvim_create_autocmd({ "BufReadPost", "BufNewFile" }, {
			pattern = "*",
			callback = function()
				-- if vim.bo.filetype ~= "NvimTree" then -- If you had problems with nvimtree.
				set_spell_settings()
			end,
		})
	end,
})
```
It took me a whole day to find these few lines of code to work.
